pkg/costmodel/allocation_helpers.go

```diff
@@ -0,0 +1,133 @@
+	PointType string
+	Key       podKey
+}
+
+// NewIntervalPoint creates and returns a new IntervalPoint instance with given parameters.
+func NewIntervalPoint(time time.Time, pointType string, key podKey) IntervalPoint {
+	return IntervalPoint{
+		Time:      time,
+		PointType: pointType,
+		Key:       key,
+	}
+}
+
+// getIntervalPointFromWindows takes a map of podKeys to windows
+// and returns a sorted list of IntervalPoints representing the
+// starts and ends of all those windows.
+func getIntervalPointsFromWindows(windows map[podKey]kubecost.Window) []IntervalPoint {
+
+	var intervals []IntervalPoint
+
+	for podKey, podInterval := range windows {
+
+		start := NewIntervalPoint(*podInterval.Start(), "start", podKey)
+		end := NewIntervalPoint(*podInterval.End(), "end", podKey)
+
+		intervals = append(intervals, []IntervalPoint{start, end}...)
+
+	}
+
+	sortIntervalPoints(intervals)
+
+	return intervals
+
+}
+
+// sortIntervalPoints sorts a list of IntervalPoints from earliest
+// to latest IntervalPoint.Time. In the case that two IntervalPoints
+// have the same time, the point with Type "start" is treated as coming
+// before the "end".
+func sortIntervalPoints(intervals []IntervalPoint) {
+	sort.Slice(intervals, func(i, j int) bool {
+		if intervals[i].Time.Equal(intervals[j].Time) {
+			return intervals[i].PointType == "start" && intervals[j].PointType == "end"
+		}
+		return intervals[i].Time.Before(intervals[j].Time)
+	})
+}
+
+// getPVCCostCoefficients gets a coefficient which represents the scale
+// factor that each PVC in a pvcIntervalMap and corresponding slice of
+// IntervalPoints intervals uses to calculate a cost for that PVC's PV.
+func getPVCCostCoefficients(intervals []IntervalPoint, pvcIntervalMap map[podKey]kubecost.Window, pvcCostCoefficientMap map[podKey][][]float64) {
+
+	// pvcCostCoefficientMap is mutated in this function. The format is
+	// such that the individual coefficient components are preserved for
+	// testing purposes.
+
+	var activePods float64
+
+	activeKeys := make(map[podKey]struct{})
+
+	// For each interval i.e. for any time a pod-PVC relation ends or starts...
+	for i := 1; i < len(intervals); i++ {
+
+		point := intervals[i]
+		prevPoint := intervals[i-1]
+
+		// intervals will always have at least two IntervalPoints (one start/end)
+		if i == 1 {
+			activePods += 1
+			activeKeys[prevPoint.Key] = struct{}{}
+		}
+
+		// If the current point happens at a later time than the previous point
+		if !point.Time.Equal(prevPoint.Time) {
+			for key := range activeKeys {
+				// Add a two-long slice of float64 representing:
+				// 1. The division of cost based on how many pods were running between those points
+				// 2. The ratio of the time between those points to the total time that pod was running
+				pvcCostCoefficientMap[key] = append(
+					pvcCostCoefficientMap[key],
+					[]float64{1.0 / activePods, point.Time.Sub(prevPoint.Time).Minutes() / pvcIntervalMap[key].Duration().Minutes()},
```

Comment:
Also, do you think it's worth `NaN` checking here? (Both values could be `NaN`, technically.)

It's obviously unlikely that `pvcIntervalMap[key].Duration().Minutes()` will be zero, but you never really know what's possible with Prometheus data, and `NaN`s tend to spread unhappiness.

Reply:
I think `NaN` checking is probably a good idea. 

I considered creating a new struct here instead of using slices, but thought it would be a little too cluttered for what is essentially a broken-down addition problem that is turned into a single number in short order. It certainly would make a lot more sense visually to be able to see something like:

```
CoefficientComponent{
	Time: point.Time.Sub(prevPoint.Time).Minutes()
	Proportion: 1.0 / activePods
}
```

so I'll try implementing it. I'd have used a tuple if they existed.

Reply:
Makes complete sense. Naming a new struct does make for an extra task, which I appreciate. But the two downsides of using a slice are:

(1) making use of that data structure in the future requires someone to go back and figure out what `[][]float64` means -- are there always a set number of entries? what do they mean? and

(2) Go will be more economical allocating memory to a well-defined struct than to a variable-length slice

Reply:
Implemented the `CoefficientComponent` type. Hopefully it's clearer.